### PR TITLE
ORD-2029: Created Alerting System Project - api service + mysql

### PR DESCRIPTION
**Pull request by GillNitzan**

### CODE REVIEW

File: tests/integration/test_alerting_system_basic.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+                    "level": 2
+                }
+            ],
+            "shop_ids_list": [
+                "ecb6ae9d-9c6c-4bcc-8ade-b6dfdd805061"
+            ]
+        }
+
+    def test_integration_basic(self):
+        time.sleep(20)
```
Comment (ido-tapingo) [inline tests/integration/test_alerting_system_basic.py]: #nowaiting
don't idle-wait for so long in your test. wait on something to happen.

File: alerting_system/api/handlers/configurations.py [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+        if not self.data.priorities:
+            return False, EMPTY_PRIORITIES_ERROR
+
+        for shop_id in self.data.shop_ids_list:
+            if not is_valid_uuid4(shop_id):
+                return False, INVALID_SHOP_ID_ERROR
+
+        return True, ""
+
+    def mark_configuration_for_deletion(self, config_id):
```
Comment (tapavel) [inline alerting_system/api/handlers/configurations.py]: I know it's a configuration, it's in the configuration class...
just mark_for_deletion

File: alerting_system/api/handlers/configurations.py [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+        session = self.session()
+        [configuration] = session.query(Configuration).filter(Configuration.id == config_id).all()
+        configuration.deleted = True
+        configuration.is_active = False
+        for priority in session.query(Priority).filter(Priority.configuration_id == configuration.id).all():
+            priority.deleted = True
+        # session.delete(configuration)
+        session.commit()
+
+    def create_configuration(self):
```
Comment (tapavel) [inline alerting_system/api/handlers/configurations.py]: just create

File: alerting_system/api/handlers/configurations.py [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+            session.add(priority)
+            session.commit()
+            priorities.append(priority)
+        return priorities
+
+    @staticmethod
+    def form_configuration_response(configuration, priorities, shop_ids_list):
+        config_dict = configuration.as_dict()
+        config_dict['shop_ids_list'] = shop_ids_list
+        config_dict['priorities'] = list(map(lambda p: p.as_dict(), priorities))
```
Comment (tapavel) [inline alerting_system/api/handlers/configurations.py]: don't use map, use list comprehensions

File: alerting_system/api/handlers/configurations.py [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+        shop_ids_list = session.query(ConfigurationShopLink.shop_id).filter(
+            ConfigurationShopLink.configuration_id == configuration.id
+        ).distinct()
+        return [r.shop_id for r in shop_ids_list]
+
+    def get_shop_ids_list_arg(self):
+        shop_ids_list = []
+        if self.is_get():
+            shop_ids = (self.get_argument('shop_ids'))
+            shop_ids_list = list(filter(lambda x: x or is_valid_uuid4(x), shop_ids.split(",")))
```
Comment (tapavel) [inline alerting_system/api/handlers/configurations.py]: don't use filter, use list comprehensions

File: alerting_system/api/handlers/errors.py [hunk 6]
```diff
@@ -0,0 +1,9 @@
+from tornado.httpclient import HTTPError
+
+EMPTY_SHOP_IDS_LIST_ERROR = "Configuration must be associated with at least one shop"
```
Comment (tapavel) [inline alerting_system/api/handlers/errors.py]: Use python enums

File: alerting_system/api/requirements/base.txt [hunk 7]
```diff
@@ -0,0 +1,10 @@
+tornado==5.0.2
+tornado-sqlalchemy==0.4.1
+raven==6.6.0
```
Comment (tapavel) [inline alerting_system/api/requirements/base.txt]: please sort by alphabetical order

File: alerting_system/models/alerts.py [hunk 8]
```diff
@@ -0,0 +1,16 @@
+"""from typing import List
+
```
Comment (tapavel) [inline alerting_system/models/alerts.py]: delete this

File: tests/integration/test_alerting_system_basic.py [hunk 9]
```diff
@@ -0,0 +1,35 @@
+import unittest
+import requests
+import time
+from bash import bash
+
+
+class AlertingSystemIntegrationBasicTestCase(unittest.TestCase):
+    def setUp(self):
+        super(AlertingSystemIntegrationBasicTestCase, self).setUp()
+        bash("cd ../../; make dev/up")
```
Comment (tapavel) [inline tests/integration/test_alerting_system_basic.py]: what's wrong with subprocess?

File: alerting_system/api/handlers/base.py [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+                data = json.loads(self.request.body)
+            except:
+                self.send_configuration_error(400, reason='JSON required')
+                return
+
+            if self.schema:
+                try:
+                    jsonschema.validate(data, self.schema)
+                except Exception as e:
+                    self.send_configuration_error(400, reason='Validation error: {}'.format(e.message))
```
Comment (tapavel) [inline alerting_system/api/handlers/base.py]: there is no message attribute to an exception

**Comment by tapavel** [comment]
please add flake8 and pylint checks

File: alerting_system/models/alerts.py (hunk 8 above)
Comment (GillNitzan) [reply alerting_system/models/alerts.py]: ack

File: alerting_system/api/requirements/base.txt (hunk 7 above)
Comment (GillNitzan) [reply alerting_system/api/requirements/base.txt]: ack

File: alerting_system/api/handlers/errors.py (hunk 6 above)
Comment (GillNitzan) [reply alerting_system/api/handlers/errors.py]: ack

File: alerting_system/api/handlers/configurations.py [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+
+        for shop_id in self.data.shop_ids_list:
+            if not is_valid_uuid4(shop_id):
+                return False, INVALID_SHOP_ID_ERROR
+
+        return True, ""
+
+    def mark_configuration_for_deletion(self, config_id):
+        session = self.session()
+        [configuration] = session.query(Configuration).filter(Configuration.id == config_id).all()
```
[the remark this answers is not in the document]
Comment (GillNitzan) [reply alerting_system/api/handlers/configurations.py]: ack

File: alerting_system/api/docker/development.compose [hunk 12]
```diff
@@ -0,0 +1,47 @@
+version: '3'
```
[the remark this answers is not in the document]
Comment (GillNitzan) [reply alerting_system/api/docker/development.compose]: ack
Changed it for better convention but actually only the default compose file is docker-compose.yml. the rest can be everything you want (when using -f flag)

File: alerting_system/api/docker/nginx-api-proxy.Dockerfile [hunk 13]
```diff
@@ -0,0 +1,3 @@
+FROM nginx:latest
+COPY development-api-proxy.conf /etc/nginx/conf.d/default.conf
```
[the remark this answers is not in the document]
Comment (GillNitzan) [reply alerting_system/api/docker/nginx-api-proxy.Dockerfile]: so please show me how :-)
I would love to remove it..

File: alerting_system/api/handlers/base.py [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+        super(BaseHandler, self).initialize()
+        self.slack_client = self.settings['slack_client']
+        self.data = ObjectDict()
+        self.schema = None
+
+    def set_default_headers(self):
+        self.set_header("Access-Control-Allow-Origin", "*")
+        self.set_header("Access-Control-Allow-Headers", "x-requested-with")
+        self.set_header('Access-Control-Allow-Methods', 'POST, GET, DELETE, OPTIONS')
+
```
[the remark this answers is not in the document]
Comment (GillNitzan) [reply alerting_system/api/handlers/base.py]: ack

**Pull request closed without merging**